kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerManager.java

```diff
@@ -274,33 +303,107 @@ public String createConsumer(String group, ConsumerInstanceConfig instanceConfig
+                // like the new consumer does. When the new consumer is available we may be
+                // able to update this to provide better feedback to the user.
+                Exception responseException = e;
+                if (!(e instanceof RestException)) {
+                  responseException = Errors.kafkaErrorException(e);
+                }
+                callback.onCompletion(null, responseException);
+              } else {
+                callback.onCompletion(records, null);
               }
-              callback.onCompletion(null, responseException);
-            } else {
-              callback.onCompletion(records, null);
             }
           }
-        }
     );
+    return executor.submit(new RunnableReadTask(new ReadTaskState(task, state, callback)));
+  }
+
+  class RunnableReadTask implements Runnable, Delayed {
+    private final ReadTaskState taskState;
+    private final KafkaRestConfig consumerConfig;
+    private final long started;
+    private final long requestExpiration;
+    // Expiration if this task is waiting, considering both the expiration of the whole task and
+    // a single backoff, if one is in progress
+    private long waitExpiration;
+
+    public RunnableReadTask(ReadTaskState taskState) {
+      this.taskState = taskState;
+      this.started = config.getTime().milliseconds();
+      this.consumerConfig = taskState.consumerState.getConfig();
+      this.requestExpiration = this.started
+              + consumerConfig.getInt(KafkaRestConfig.CONSUMER_REQUEST_TIMEOUT_MS_CONFIG);
+      this.waitExpiration = 0;
+    }
+
+    @Override
+    public void run() {
+      try {
+        log.trace("Executing consumer read task ({})", taskState.task);
+        long backoffTime = config.getTime().milliseconds()
+                + consumerConfig.getInt(KafkaRestConfig.CONSUMER_ITERATOR_BACKOFF_MS_CONFIG);
+
+        taskState.task.doPartialRead();
+        if (!taskState.task.isDone()) {
+          waitExpiration = Math.min(backoffTime, requestExpiration);
+          if (delayedTasks.get(this) != null && delayedTasks.get(this)) {
+            // task is already enqueued
+            return;
+          }
+
+          // add to delayedReadTasks so the scheduler thread can re-schedule another partial read
+          delayedTasks.put(this, true);
+          delayedReadTasks.add(this);
+        } else {
+          log.trace("Finished executing consumer read task ({})", taskState.task);
+        }
+      } catch (Exception e) {
+        log.error("Failed to read records consumer "
+                        + taskState.consumerState.getId().toString(),
+                e);
+        Exception responseException = e;
+        if (!(e instanceof RestException)) {
+          responseException = Errors.kafkaErrorException(e);
+        }
+        taskState.callback.onCompletion(null, responseException);
+      }
+    }
+
+    @Override
+    public long getDelay(TimeUnit unit) {
```

Comment:
Should we be ignoring the time unit? This might be used internally in `DelayQueue`, so we had better not I think.

Reply:
It uses `TimeUnit.NANOSECONDS`. From the docs:
```
.Expiration occurs when an element's getDelay(TimeUnit.NANOSECONDS) method returns a value less than or equal to zero.
```
Since we keep track in milliseconds and we keep it consistent (always return millis), I had thought it's not important.

If we would not ignore the parameter, it would make sense to track expirations in nanoseconds  or make our variables `double`, since we currently can't go more fine-grained than 1ms. I think there's no use - wdyt?